src/Microsoft.VisualStudio.LanguageServices.Razor/ProjectSystem/FallbackRazorProjectHost.cs

```diff
@@ -129,6 +161,97 @@ protected virtual Version GetAssemblyVersion(string filePath)
             return ReadAssemblyVersion(filePath);
         }
 
+        // Internal for testing
+        internal HostDocument[] GetCurrentDocuments(IProjectSubscriptionUpdate update)
+        {
+            var documents = new List<HostDocument>();
+
+            // Content Razor files
+            if (update.CurrentState.TryGetValue(ContentItem.SchemaName, out var rule))
+            {
+                foreach (var kvp in rule.Items)
+                {
+                    if (TryGetRazorDocument(kvp.Value, out var document))
+                    {
+                        documents.Add(document);
+                    }
+                }
+            }
+
+            // None Razor files, these are typically included when a user links a file in Visual Studio.
+            if (update.CurrentState.TryGetValue(NoneItem.SchemaName, out var nonRule))
+            {
+                foreach (var kvp in nonRule.Items)
+                {
+                    if (TryGetRazorDocument(kvp.Value, out var document))
+                    {
+                        documents.Add(document);
+                    }
+                }
+            }
+
+            return documents.ToArray();
```

Comment:
Do we need to remove duplicates?

Reply:
That's handled by the underlying system. We eagerly feed duplicate info (also in the default host) down into the `ProjectSnapshotManager`/`RazorProjectHostBase`. Those ensure they ignore duplicates.